### PR TITLE
[flutter_tools] move archive extractor to a separate library and make archive extraction failure a tool exit

### LINKED ISSUE #72386: Consider throwToolExit when archive extraction fails
The code here:

https://github.com/flutter/flutter/blob/34d4c7fe018f74c63e60b9f442fec6973c48caca/packages/flutter_tools/lib/src/cache.dart#L1794

`rethrows` an exception if the last download/extract retry fails. Instead we should consider it a network failure since the bits are likely corrupt, and throw a ToolExit as in the case above:

https://github.com/flutter/flutter/blob/34d4c7fe018f74c63e60b9f442fec6973c48caca/packages/flutter_tools/lib/src/cache.dart#L1764

/cc @jonahwilliams who was working in this area recently.

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Moves archive extractor to a separate library, to reduce the size of the far too large cache.dart library. Additionally, makes archive extractor a final and set in the constructor, taking advantage of the previous change I made to initialize the flutterRoot earlier.

Makes archive extraction failure a tool exit instead of rethrowing. Fixes #72386

### CODE REVIEW

File: packages/flutter_tools/lib/src/artifact_updater.dart [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+      } on Exception catch (err) {
+        retries -= 1;
+        if (retries == 0) {
+          throwToolExit(
+            'Flutter could not download and/or extract $url. Ensure you have '
+            'network connectivity and all of the required dependencies listed at'
+            'flutter.dev/setup.\nThe original exception was: $err.'
+          );
+        }
+        _deleteIgnoringErrors(tempFile);
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/artifact_updater.dart:155]: This is the only change to this library

File: packages/flutter_tools/test/commands.shard/hermetic/install_test.dart [hunk 2]
```diff
@@ -33,7 +33,7 @@ void main() {
 
       await createTestCommandRunner(command).run(<String>['install']);
     }, overrides: <Type, Generator>{
-      Cache: () => Cache.test(),
+      Cache: () => Cache.test(processManager: FakeProcessManager.any()),
```
Comment (zanderso) [inline packages/flutter_tools/test/commands.shard/hermetic/install_test.dart:36]: Does Cache.test() ever need a different ProcessManager?

Comment (jonahwilliams) [reply packages/flutter_tools/test/commands.shard/hermetic/install_test.dart:36]: It shouldn't, the process manager is only used for chmod.

**Comment by jonahwilliams** [comment]
still looking into why this is failing

**Comment by jonahwilliams** [comment]
needs more work

**Pull request closed without merging**